### ISSUE TITLE
Object.keys() Does Not Work Properly

### ISSUE THREAD

**Issue by c1moore** [Automattic/mongoose#5586]
This is a **bug**.  I am using Mongoose version **4.11.8**.

The best way to explain this bug is with an example, so here is a minimally reproducible script:

```javascript
'use strict';

const mongoose = require('mongoose');
const Schema = mongoose.Schema;

mongoose.Promise = require('bluebird');

var Test;

function endProcess(error) {
  error && console.error(error);

  Test.remove({});

  process.exit();
};

mongoose.connect('mongodb://localhost/test', {
  useMongoClient: true
}).then(
  function() {
    var test;

    const TestSchema = new Schema({
      A1: String,
      A2: String,
      A3: String,
      A4: {
        B1: String,
        B2: {
          C1: String
        }
      }
    });

    Test = mongoose.model('Test', TestSchema);

    test = new Test();

    console.log(Object.keys(test.A4.B2));
  },
  function(error) {
    console.error(error);

    process.exit();
  }
).then(endProcess, endProcess);
```

## Expected Behavior

I expect `console.log(Object.keys(test.A4.B2));` to print the following

`[ 'C1' ]`

## Actual Behavior

Instead the following is printed

`[ 'A1', 'A2', 'A3', 'A4', '_id', '__v', 'id', 'C1']`

Notice the keys for the top level AND the actual Object I am calling `Object.keys()` on are printed, but NOT the keys for `A4`.

**However,** if I instead execute `console.log(Object.keys(test.A4));` the expected results are printed (i.e. `[ 'B2', 'B1' ]`), which suggests that is is indeed a bug and not expected behavior.

**Comment by varunjayaraman** [comment]
mongoose objects have a lot of getters and setters and weird internal state, i wouldn't use `Object.keys` on a mongoose object

**Comment by c1moore** [comment]
It works for the first and second level objects, why not further down the structure?  What would an alternative be when I need to iterate over the keys of a Mixed field type?

**Comment by varunjayaraman** [comment]
call `.toObject()` and then iterate over the keys

**Comment by c1moore** [comment]
It seems to me like that is more of a monkey patch.  `toObject()` has its own set of problems:

- `toObject()` may not return the exact version of the current mongoose model.  We can transform the document using the `toObject.getters`, `toObject.virtuals`, and `toObject.transform` schema options.  This is the case for me as I'm using `toObject.transform` to transform some of the fields I'm iterating over.
- Depending on the implementation, when navigating over a document, a single subdocument can be converted into an object up to _d_ times, where _d_ is the depth of the subdocument in relation to the main document.  This can happen, for example, when you need to work with the actual document, not an Object version of the document.  In the example above, `C1` could be converted into an Object up to 3 times (again depending on the implementation).
- It just seems dirty converting one thing into something else just to perform a simple operation on it that should be handled by the original object.

`Object.keys()` works on the original document _and_ the second-level document.  This means Mongoose is smart enough to hide all that "weird internal state" (probably by using a Proxy), so that shouldn't factor into whether or not this bug should be fixed.  In fact, it suggests that `Object.keys()` is meant to work despite "a lot of getters and setters and weird internal state".  In fact, `Object.keys()` only prints the keys on the document while `console.log()` prints the _entire_ object, including that weird internal state.  Once Mongoose reaches a depth of 3, the _behavior of `Object.keys()` changes_.  [That's not good](https://books.google.com/books?id=LpVCAwAAQBAJ&pg=PA102&lpg=PA102&dq=code+complete+semantic+coupling&source=bl&ots=GHHohhCgOK&sig=kVgm8tZ2YtzV4CdaePp8Z-W583Q&hl=en&sa=X&ved=0ahUKEwi6yuH7u4HWAhVl64MKHYEfB4gQ6AEIQDAD#v=onepage&q=code%20complete%20semantic%20coupling&f=false).

**Comment by c1moore** [comment]
Also, this logic appears in the `toObject()` method, so calling `toObject()` is not really a solution for me.

**Comment by varunjayaraman** [comment]
looping @vkarpov15 in. From personal experience, I've done it by accessing `doc._doc` or by calling `.toObject()`. If you know the nested thing is going to be a mongoose doc, you could do `Object.keys(doc.nestedDoc._doc)` but other than that I would agree mongoose docs can be a bit wonky

**Comment by vkarpov15** [comment]
This is a duplicate of #5470, we have a PR outstanding to fix this: https://github.com/Automattic/mongoose/pull/5546 but waiting for 4.12 to release it because it's a little too substantial a change to put in a point release.

**Comment by vkarpov15** [comment]
Fixed by #5546
